### PR TITLE
Spinboxes now correct when rounding.

### LINKED ISSUE #5298: [GNU/LINUX] Grid snapping and angle snapping not exact digits
**Describe the bug**
Enabled grid snapping and angle snapping digits is `0.4` sometimes not exact digits `0.39999913` this is a bug

**Screenshots/Video**

https://user-images.githubusercontent.com/17591539/140290148-3926c865-36a8-4b4a-a311-902d41173554.mp4

https://user-images.githubusercontent.com/17591539/140290177-a9787659-a64c-4e5a-b2a1-76c8800c16cb.mp4

**Found in Branch**
stabilization/2110 5a0b6314e4da5446e8888bff7344203e73e09bd8

**Desktop/Device (please complete the following information):**
 - Device: PC
 - OS: Xubuntu
 - Version: 20.04.03 LTS
 -  CPU: AMD 3 3100
 - GPU: AMD 5500 XT [Mesa 21.0.3]
 - Memory: 16GB

### PR DESCRIPTION
**Pull request by jjjoness**
Signed-off-by: John Jones-Steele <82226755+jjjoness@users.noreply.github.com>
Fixes #5298
spinbox

### CODE REVIEW

**Review by AMZN-daimini — APPROVED** [review]
Could this also be used on sliders to ensure the input and the tooltip display the same value?

**Comment by jjjoness** [comment]
@AMZN-daimini I'll have a look and see. I'll create a PR if it can.

File: Code/Framework/AzQtComponents/AzQtComponents/Utilities/Conversions.cpp [hunk 1]
```diff
@@ -39,15 +39,15 @@ namespace AzQtComponents
...
         const QChar decimalPoint = locale.decimalPoint();
         const QChar zeroDigit = locale.zeroDigit();
         const int numToStringDecimals = AZStd::max(numDecimals, 20);
 
-        // We want to truncate, not round. toString will round, so we add extra decimal places to the formatting
-        // so we can remove the last values
-        QString retValue = locale.toString(value, 'f', (numDecimals > 0) ? numToStringDecimals : 0);
+        // If we want to truncate, not round, we add extra decimal places to the formatting
+        // so we can remove the last values otherwise we allow rounding
+        QString retValue = locale.toString(value, 'f', (numDecimals > 0) ? (truncate ? numToStringDecimals : numDecimals) : 0);
```
Comment (hultonha) [inline Code/Framework/AzQtComponents/AzQtComponents/Utilities/Conversions.cpp:50]: suggestion: Could we replace the nested ternary statement here with an if statement (or immediately invoked lambda) It just gets pretty hard to parse as it is right now

File: Code/Framework/AzQtComponents/AzQtComponents/Components/Widgets/SpinBox.cpp [hunk 2]
```diff
@@ -1460,7 +1460,7 @@ QString DoubleSpinBox::stringValue(double value, bool truncated) const
         numDecimals = 0;
     }
 
-    return toString(value, numDecimals, locale(), isGroupSeparatorShown());
+    return toString(value, numDecimals, locale(), isGroupSeparatorShown(), false);
```
Comment (hultonha) [inline Code/Framework/AzQtComponents/AzQtComponents/Components/Widgets/SpinBox.cpp:1463]: question: This is likely intended it just seems odd on first glance that we don't pass the `truncated` parameter on line `1450` through to this call here. Are they different kind of truncation? Could the logic here be simplified if we just pass the bool through to `toString`?

File: Code/Framework/AzQtComponents/AzQtComponents/Utilities/Conversions.cpp [hunk 3]
```diff
@@ -39,15 +39,15 @@ namespace AzQtComponents
         return AZ::Color(static_cast<float>(rgb.redF()), static_cast<float>(rgb.greenF()), static_cast<float>(rgb.blueF()), static_cast<float>(rgb.alphaF()));
     }
 
-    QString toString(double value, int numDecimals, const QLocale& locale, bool showGroupSeparator)
+    QString toString(double value, int numDecimals, const QLocale& locale, bool showGroupSeparator, bool truncate)
```
Comment (hultonha) [inline Code/Framework/AzQtComponents/AzQtComponents/Utilities/Conversions.cpp:42]: suggestion: It might be useful to have a couple of quick sanity tests for this function to ensure we get the expected value. They could be really simple functions with no fixture, there should hopefully be a place for them in `AzQtComponents` I'd have thought. Could be a separate PR but would be great to capture the expected behavior of this function given various inputs

**Review by hultonha — COMMENTED** [review]
Overall looks good I just have a small number of questions/comments. I can approve after that as nothing is super blocking but might be worth doing another pass to simplify things if possible.

File: Code/Framework/AzQtComponents/AzQtComponents/Utilities/Conversions.cpp (hunk 1 above)
Comment (jjjoness) [reply Code/Framework/AzQtComponents/AzQtComponents/Utilities/Conversions.cpp:50]: Done.

File: Code/Framework/AzQtComponents/AzQtComponents/Components/Widgets/SpinBox.cpp (hunk 2 above)
Comment (jjjoness) [reply Code/Framework/AzQtComponents/AzQtComponents/Components/Widgets/SpinBox.cpp:1463]: They are different. The value passed in means to either truncate to the number of decimals of to the numberofDisplayDecimals. I've reworked the logic so the term to pass is the round bool.

File: Code/Framework/AzQtComponents/AzQtComponents/Utilities/Conversions.cpp (hunk 3 above)
Comment (jjjoness) [reply Code/Framework/AzQtComponents/AzQtComponents/Utilities/Conversions.cpp:42]: Done😀

File: Code/Framework/AzQtComponents/AzQtComponents/Utilities/Conversions.cpp (hunk 1 above)
Comment (hultonha) [reply Code/Framework/AzQtComponents/AzQtComponents/Utilities/Conversions.cpp]: Legend, thanks 🙂

File: Code/Framework/AzQtComponents/AzQtComponents/Components/Widgets/SpinBox.cpp (hunk 2 above)
Comment (hultonha) [reply Code/Framework/AzQtComponents/AzQtComponents/Components/Widgets/SpinBox.cpp]: Okay nice one, and thanks for the explanation 👍

File: Code/Framework/AzQtComponents/AzQtComponents/Tests/AzQtComponentTests.cpp [hunk 4]
```diff
@@ -34,3 +36,17 @@ class AzQtComponentsTestEnvironment : public AZ::Test::ITestEnvironment
...
+    QString result = AzQtComponents::toString(testVal, 3, QLocale(), false, false);
+    EXPECT_TRUE(result == "1.239");
+}
+
+TEST(AzQtComponents, ToStringReturnsRoundedString)
+{
+    double testVal = 1.2399999;
+    QString result = AzQtComponents::toString(testVal, 3, QLocale(), false, true);
+    EXPECT_TRUE(result == "1.24");
+}
```
Comment (hultonha) [inline Code/Framework/AzQtComponents/AzQtComponents/Tests/AzQtComponentTests.cpp:52]: Excellent! 😄 Thanks for adding these 🙂

**Review by hultonha — APPROVED** [review]
LGTM! 😄 Thanks for making the various updates, much appreciated 👍
